test/test_nn.py

```diff
@@ -61,7 +61,7 @@
 
 dtype2prec = {torch.float: 1e-5,
               torch.double: 1e-5,
-              torch.half: 1e-2}
```

Comment:
didn't review other parts. but does this pr make the computation less precise?

Reply:
Actually, it is more precise. 
But now that I have also implemented the non-bag version, the reduction in the comparison sensitivity isn't need.